### DISCUSSION TITLE
Error on ARM64 Windows 11 when executing dotnet.exe

### DISCUSSION THREAD

**Discussion by christophwille** [Tyrrrz/CliWrap#249] [Q&A] [answered]
This test https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler.Tests/CorrectnessTestRunner.cs#L38 calls the underlying dotnet.exe compilation that is wrapped with CliWrap https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler.Tests/Helpers/Tester.cs#L135. That has worked and still is working great on x64.

On ARM64 I get the following:

```
 AllFilesHaveTests
   Source: CorrectnessTestRunner.cs line 38
   Duration: 25.8 sec

  Message: 
OneTimeSetUp: CliWrap.Exceptions.CommandExecutionException : Command execution failed because the underlying process (dotnet.exe#52568) returned a non-zero exit code (-2147450735).

Command:
dotnet.exe build Z:\GitWorkspace\ILSpy\ICSharpCode.Decompiler.Tests\bin\Debug\net8.0-windows\win-x64\../../../../../ICSharpCode.Decompiler.TestRunner/ICSharpCode.Decompiler.TestRunner.csproj -r win-x86 -c Debug --self-contained

You can suppress this validation by calling `WithValidation(CommandResultValidation.None)` on the command.

  Stack Trace: 
Command.ExecuteAsync(ProcessEx process, CancellationToken forcefulCancellationToken, CancellationToken gracefulCancellationToken) line 271
Command.ExecuteAsync(ProcessEx process, CancellationToken forcefulCancellationToken, CancellationToken gracefulCancellationToken) line 287
Command.ExecuteAsync(ProcessEx process, CancellationToken forcefulCancellationToken, CancellationToken gracefulCancellationToken) line 287
Command.ExecuteAsync(ProcessEx process, CancellationToken forcefulCancellationToken, CancellationToken gracefulCancellationToken) line 287
Tester.BuildTestRunner(String runtime, String config) line 145
Tester.Initialize() line 136
ToolsetSetup.RunBeforeAnyTests() line 55
GenericAdapter`1.BlockUntilCompleted()
NoMessagePumpStrategy.WaitForCompletion(AwaitAdapter awaiter)
AsyncToSyncAdapter.Await[TResult](Func`1 invoke)
AsyncToSyncAdapter.Await(Func`1 invoke)
SetUpTearDownItem.RunSetUpOrTearDownMethod(TestExecutionContext context, IMethodInfo method)
SetUpTearDownItem.RunSetUp(TestExecutionContext context)
<.ctor>b__0(TestExecutionContext context)
BeforeTestCommand.Execute(TestExecutionContext context)
BeforeTestCommand.Execute(TestExecutionContext context)
CompositeWorkItem.PerformOneTimeSetUp()
```

If I take the command and paste it in Terminal, even though I build win-x86, it runs just fine with no complaints whatsover. CliWrap is latest, .NET 8.0 latest from yesterday's Patch Tuesday, same for VS. Any ideas (and possibly how to debug?)?

**Comment by Tyrrrz** [comment]
Try running it with `ExecuteBufferedAsync()` to capture the stderr in the error message

**Comment by christophwille** [comment]
It still throws with ExecuteBufferedAsync (and thus result = is null), but the exception text in the VS debugger is more helpful:

```
Command execution failed, see the inner exception for details.

Standard error:
The command could not be loaded, possibly because:
  * You intended to execute a .NET application:
      The application 'build' does not exist.
  * You intended to execute a .NET SDK command:
      A compatible .NET SDK was not found.

Requested SDK version: 8.0.100
global.json file: Z:\GitWorkspace\ILSpy\global.json

Installed SDKs:
No .NET SDKs were found.

Install the [8.0.100] .NET SDK or update [Z:\GitWorkspace\ILSpy\global.json] to match an installed SDK.

Download a .NET SDK:
https://aka.ms/dotnet/download

Learn about SDK resolution:
https://aka.ms/dotnet/sdk-not-found
```

The global.json

```
{
  "sdk": {
    "version": "8.0.100",
    "rollForward": "major",
    "allowPrerelease": true
  }
} 
```

I am racking my mind why that would behave that way... setting to .303 doesn't help. PS as well as command both execute dotnet.exe nicely, so those two find dotnet correctly. Q then is why the non-interactive variation with CliWrap fails... almost looks like PATH isn't picked up.

**Reply by Tyrrrz** [reply]
> but the exception text in the VS debugger is more helpful:

Yeah that was the goal. It captures stderr so shows more info in the error message.

**Comment by christophwille** [comment]
Right, `await Cli.Wrap("c:\\program files\\dotnet\\dotnet.exe")` instead of `await Cli.Wrap("dotnet.exe")` works. But why would an interactive prompt properly pick up the system env var PATH and CliWrap wouldn't (on ARM64 compared to x64)?

**Comment by christophwille** [comment]
`Environment.GetEnvironmentVariable("PATH")` however contains the expected env var (in my code before the CliWrap call, which should be good enough). Is there any custom path logic inside CliWrap?

**Reply by Tyrrrz** [reply]
> Is there any custom path logic inside CliWrap?

Yes, but for `exe` files it should be the same as the system resolution logic.

https://github.com/Tyrrrz/CliWrap/blob/9e4e6bfc699e8a1771c08d58d6e9d3b60ebbab60/CliWrap/Command.Execution.cs#L17-L71

**Comment by christophwille** [comment, marked as answer]
I let that error message simmer a little bit - and then it dawned on me: there is an x64 RID on that test project, and thus it was starting the CLI in that process mode as well, and of course there is no x64 net8.0 SDK on that machine, only the ARM64 version. Setting 

`<RuntimeIdentifier>win-arm64</RuntimeIdentifier>`

made the unmodified code work right away. Although I am still surprised that the prior full-path-variant worked....

**Reply by Tyrrrz** [reply]
I know some environment variables may have different values if you run an app in emulation, such as x86 on x64 or x64 on arm64, etc. But if you checked `PATH` and it's right, then I'm not sure exactly what causes it. Maybe the .NET Host Resolver detects the target architecture somehow and resolves to an x64-based runtime.
